lib/proxy.js

```diff
@@ -806,34 +820,39 @@ Proxy.prototype._onHttpServerRequest = function(isSSL, clientToProxyRequest, pro
         if (err) {
           return self._onError('ON_REQUESTHEADERS_ERROR', ctx, err);
         }
-        return makeProxyToServerRequest();
+        if(self.cache) {
+          const key = self.cache.keyFn(ctx.clientToProxyRequest);
+          self.cache.redisClient.get(key, (err, result) => {
+            if(err || !result) {
+              return makeProxyToServerRequest();
+            }
+            const cachedResponse = JSON.parse(result);
+            const responseLike = new ResponseLike(
+                cachedResponse.statusCode,
+                cachedResponse.headers,
+                Buffer.from(cachedResponse.body),
+                cachedResponse.url
+            );
+            responseLike.fromCache = true;
+            return proxyToServerRequestComplete(responseLike);
+          });
+        }
+        else return makeProxyToServerRequest();
       });
     });
   }
 
   function makeProxyToServerRequest() {
     var proto = ctx.isSSL ? https : http;
-
-    function onRequest(request) {
-      ctx.proxyToServerRequest = request;
-      ctx.proxyToServerRequest.on('error', self._onError.bind(self, 'PROXY_TO_SERVER_REQUEST_ERROR', ctx));
-      ctx.requestFilters.push(new ProxyFinalRequestFilter(self, ctx));
-      var prevRequestPipeElem = ctx.clientToProxyRequest;
-      ctx.requestFilters.forEach(function (filter) {
-          filter.on('error', self._onError.bind(self, 'REQUEST_FILTER_ERROR', ctx));
-          prevRequestPipeElem = prevRequestPipeElem.pipe(filter);
-      });
-      ctx.clientToProxyRequest.resume();
-    }
-
-    const cacheableRequest = ctx.isSSL ? self.httpsCacheableRequest : self.httpCacheableRequest;
-    const requestFunction = self.cacheRequests ? cacheableRequest : proto.request;
-    const requestOpts = { ...self.cacheOpts, ...ctx.proxyToServerRequestOptions };
-    const cacheableOrNormalResult = requestFunction(requestOpts, proxyToServerRequestComplete);
-    if (self.cacheRequests) {
-      cacheableOrNormalResult.on('request', onRequest);
-      cacheableOrNormalResult.on('error', self._onError.bind(self, 'CACHE_REQUEST_ERROR', ctx));
-    } else onRequest(cacheableOrNormalResult);
+    ctx.proxyToServerRequest = proto.request(ctx.proxyToServerRequestOptions, proxyToServerRequestComplete);
+    ctx.proxyToServerRequest.on('error', self._onError.bind(self, 'PROXY_TO_SERVER_REQUEST_ERROR', ctx));
+    ctx.requestFilters.push(new ProxyFinalRequestFilter(self, ctx));
+    var prevRequestPipeElem = ctx.clientToProxyRequest;
+    ctx.requestFilters.forEach(function(filter) {
```

Comment:
```suggestion
    ctx.requestFilters.forEach((filter) => {
```

Reply:
That's not my code, I'd rather keep it as is for consistency. The changes you see are caused by the revert I've done to my previous changes.

Reply:
OK